File: dingco/2nd_week/2_extra2_is_existing_target_number_binary.swift
let findingTarget = 23
let findingNumbers = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]

func isExistingTargetNumberBinary(target: Int, from array: [Int]) -> Bool {
    var midIndex = array.count / 2
    var findingStart = 0
    var findingEnd = array.count - 1

    while findingStart <= findingEnd {
        let mid = array[midIndex]

        // print("findingStart \(findingStart)")
        // print("findingEnd \(findingEnd)")
        // print("midIndex \(midIndex)\n")
        // print("mid \(mid)")

        if mid == target {
            return true
        } else if mid < target {
            findingStart = midIndex + 1 // 인덱스!!! mid가 아니라 midIndex임
        } else if mid > target {
            findingEnd = midIndex - 1
        }

        midIndex = (findingStart + findingEnd) / 2
    }

    return false
}

let result = isExistingTargetNumberBinary(target: findingTarget, from: findingNumbers)
print(result)

File: dingco/2nd_week/2_extra2_is_existing_target_number_binary.py
finding_target = 23
finding_numbers = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]

def is_existing_target_number_binary(target, array):
    if not array:
        return False

    mid_index = len(array) // 2
    mid = array[mid_index]
    if mid == target:
        return True
    elif mid < target:
        return is_existing_target_number_binary(target, array[mid_index+1:])
    elif mid > target:
        return is_existing_target_number_binary(target, array[:mid_index])


result = is_existing_target_number_binary(finding_target, finding_numbers)
print(result)
